core/trie/proof.go

```diff
@@ -133,6 +133,163 @@ func transformNode(tri *Trie, parentKey *Key, sNode StorageNode) (*Edge, *Binary
 	return edge, binary, nil
 }
 
+func traverseNodesAndAppendToPath(path *[]ProofNode, currNode *ProofNode, nodeHashes map[string]ProofNode) error {
+	whileLoop := true
+	for whileLoop {
+		switch {
+		case currNode.Binary != nil:
+			expectedLeftHash := currNode.Binary.LeftHash.String()
+			expectedRightHash := currNode.Binary.RightHash.String()
+
+			nodeLeft, leftExist := nodeHashes[expectedLeftHash]
+			nodeRight, rightExist := nodeHashes[expectedRightHash]
+			if leftExist && rightExist {
+				return errors.New("unexpected error")
+			} else if leftExist {
+				currNode = &nodeLeft
+				*path = append(*path, *currNode)
+			} else if rightExist {
+				currNode = &nodeRight
+				*path = append(*path, *currNode)
+			} else {
+				whileLoop = false
+			}
+		case currNode.Edge != nil:
+			righNode, exists := nodeHashes[currNode.Edge.Child.String()]
+			if exists {
+				currNode = &righNode
+				*path = append(*path, *currNode)
+			} else {
+				whileLoop = false
+			}
+		}
+	}
+
+	return nil
+}
+
+// Remove duplicates and merges proof paths into a single path
+func MergeProofPaths(leftPath, rightPath []ProofNode, hash hashFunc) ([]ProofNode, *felt.Felt, error) {
+	merged := []ProofNode{}
+	minLen := min(len(leftPath), len(rightPath))
+
+	if len(leftPath) == 0 || len(rightPath) == 0 {
+		return merged, nil, errors.New("empty proof paths")
+	}
+
+	if !leftPath[0].Hash(hash).Equal(rightPath[0].Hash(hash)) {
+		return merged, nil, errors.New("roots of the proof paths are different")
+	}
+
+	rootHash := leftPath[0].Hash(hash)
+
+	// Get duplicates and insert by one
+	i := 0
+	for i = 0; i < minLen; i++ {
+		leftNode := leftPath[i]
+		rightNode := rightPath[i]
+
+		if leftNode.Hash(hash).Equal(rightNode.Hash(hash)) {
+			merged = append(merged, leftNode)
+		} else {
+			break
+		}
+	}
+
+	// Add rest of the nodes one from left and one from right
+	// until we reach the end of the shortest path
+	for ; i < minLen; i++ {
+		merged = append(merged, leftPath[i], rightPath[i])
+	}
+
+	// Add the rest of the nodes from the longest path
+	if len(leftPath) > minLen {
+		merged = append(merged, leftPath[i:]...)
+	} else if len(rightPath) > minLen {
+		merged = append(merged, rightPath[i:]...)
+	}
+
+	return merged, rootHash, nil
+}
+
+func SplitProofPath(mergedPath []ProofNode, rootHash *felt.Felt, hash hashFunc) ([]ProofNode, []ProofNode, error) {
+	leftPath := []ProofNode{}
+	rightPath := []ProofNode{}
+
+	if len(mergedPath) == 0 {
+		return leftPath, rightPath, nil
+	}
+
+	nodeHashes := make(map[string]ProofNode)
```

Comment:
I think felt (result of hash) should work in map as well?
```suggestion
	nodeHashes := make(map[felt.Felt]ProofNode)
```

Reply:
We are using felt.Felt with the new commit